### ISSUE TITLE
Feature request for O(1) access to node neighbors in GraphMap

### ISSUE THREAD

**Issue by aneksteind** [petgraph/petgraph#345]
### Summary
Currently (I believe) checking to see if a node is present in the neighbor collection of another is O(|V|) access in the worst case. It should be possible to improve this to O(1).

### Motivation

What does adding this feature enable that isn't possible without it? Does it
improve performance? Turn dynamic checks into statically enforced properties?
Explain why this feature should be implemented.

Immediate Benefits:

The [`remove_single_edge`](https://github.com/petgraph/petgraph/blob/master/src/graphmap.rs#L268) function has a worst case removal of O(|V|) because of the search for a node's index; having O(1) access to neighbors of a node would reduce this operation to constant time.

Future benefits and applications:

There are a few algorithmic applications I have in mind, one of which is related to graph minors and graph contraction. In costed graph contractions, it's useful to union two neighbor sets. To do a union without iterating over the neighbors first to build a hash set, this feature would be useful.

### Details

* What code needs to change? New traits? Which modules?

`src/graphmap.rs`

The proposed change is to replace:

```rust
pub struct GraphMap<N, E, Ty> {
    nodes: IndexMap<N, Vec<(N, CompactDirection)>>,
    edges: IndexMap<(N, N), E>,
    ty: PhantomData<Ty>,
}
```

with

```rust
pub struct GraphMap<N, E, Ty> {
    nodes: IndexMap<N, IndexMap<N, CompactDirection>>,
    edges: IndexMap<(N, N), E>,
    ty: PhantomData<Ty>,
}
```

and remedy the downstream affects.

* Is this a new graph algorithm? Include links to papers, Wikipedia, or other
  implementations of this algorithm that exist.

N/A

* Are you willing to implement this yourself? Mentor someone else and help them
  implement it?

I am willing to implement this myself or mentor another

**Comment by bluss** [comment]
It might make sense as an incremental improvement, but I'd rather direct focus towards rewriting GraphMap #239 #208

**Comment by aneksteind** [comment]
@bluss what can I do to help with that. I noticed there was a call for help some time ago and see some ideas on one of those linked issues, but nothing set in stone as mentioned.

**Comment by bluss** [comment]
`GraphMap::contains_edge` is already O(1), from the issue description, one might think it doesn't exist :)

I don't know aneksteind, I don't have much time for petgraph. That means help is certainly needed.

**Comment by aneksteind** [comment]
> I don't have much time for petgraph. That means help is certainly needed.

Fair enough, is that a directive to simply begin rewriting it and submit a PR, or is a little more back and forth between contributors and the team desired?

**Comment by bluss** [comment]
Best talk with contributors IMO 🙂
